### PR TITLE
Add Chaincode Events to Transfer and Approve

### PR DESCRIPTION
**Pull request by jcastrence**
#### Type of change
- New feature
#### Description
- Transfer and Approve functions of token_contract.go now emit chaincode events.
#### Related issues
- [FAB-18275](https://jira.hyperledger.org/browse/FAB-18275)

### CODE REVIEW

File: token-account-based/chaincode-go/chaincode/token_contract.go [hunk 1]
```diff
@@ -101,6 +102,14 @@ func (s *SmartContract) Transfer(ctx contractapi.TransactionContextInterface, re
 		return fmt.Errorf("failed to transfer: %v", err)
 	}
 
+	// Emit the Transfer event
+	var transferEvent = map[string]string{"from": clientID, "to": recipient, "value": strconv.Itoa(amount)}
```
Comment (btl5037) [inline token-account-based/chaincode-go/chaincode/token_contract.go:106]: nit: Objects with known values are generally defined using `:=` unless the object is in the global scope
```
transferEvent := map[string]string{"from": clientID, "to": recipient, "value": strconv.Itoa(amount)}
```

File: token-account-based/chaincode-go/chaincode/token_contract.go [hunk 2]
```diff
@@ -200,6 +209,14 @@ func (s *SmartContract) Approve(ctx contractapi.TransactionContextInterface, spe
 		return fmt.Errorf("failed to update state of smart contract for key %s: %v", allowanceKey, err)
 	}
 
+	// Emit the Approval event
+	var approvalEvent = map[string]string{"owner": owner, "spender": spender, "value": strconv.Itoa(value)}
```
Comment (btl5037) [inline token-account-based/chaincode-go/chaincode/token_contract.go:213]: nit: Objects with known values are generally defined using `:=` unless the object is in the global scope
```
approvalEvent := map[string]string{"owner": owner, "spender": spender, "value": strconv.Itoa(value)}
```

File: token-account-based/chaincode-go/chaincode/token_contract.go (hunk 1 above)
Comment (jcastrence) [reply token-account-based/chaincode-go/chaincode/token_contract.go:106]: 👌🏼

File: token-account-based/chaincode-go/chaincode/token_contract.go [hunk 3]
```diff
@@ -200,6 +209,14 @@ func (s *SmartContract) Approve(ctx contractapi.TransactionContextInterface, spe
...
+	approvalEventJSON, err := json.Marshal(approvalEvent)
+	if err != nil {
+		return fmt.Errorf("failed to obtain JSON encoding: %v", err)
+	}
+	ctx.GetStub().SetEvent("Approval", approvalEventJSON)
```
[the remark this answers is not in the document]
Comment (jcastrence) [reply token-account-based/chaincode-go/chaincode/token_contract.go:218]: Nice catch

**Review by btl5037 — APPROVED** [review]

**Comment by btl5037** [comment]
Thanks @jcastrence looks good now

**Review by nikhil550 — APPROVED** [review]
LGTM

**Comment by nikhil550** [comment]
We can now probably update the tutorial since many of the ERC20 functions are now in the golang contract as well.

**Comment by denyeart** [comment]
@jcastrence In a follow-on PR, please emit the Tranfser event in TransferFrom() as well.

**Pull request merged**